### PR TITLE
Wire SiTU-GLU (Kimi K3) into the routed expert kernel

### LINKED ISSUE #51351: Kimi-K3 routed expert bring-up
Current implementation is `output = self.down_proj(self.act_fn(self.gate_proj(x)) * self.up_proj(x))`.
While Kimi K3 implementation for routed expert with SiTU is             
```
gate_up = torch.cat([self.gate_proj(hidden_states), self.up_proj(hidden_states)], dim=-1)
current_hidden_states = self.act_fn(gate_up)
output = self.down_proj(current_hidden_states)
```

### PR DESCRIPTION
**Pull request by kgrujcic-tt**
Wires the SiTU-GLU activation (Kimi K3) into the fused routed-expert FFN kernel,
following the same pattern the gpt-oss/MiniMax-M3 SwiGLU-OAI activation uses.

Targets `kgrujcic/51350-situ` (#52448), which adds the `situ_glu` SFPU op this builds on.
Closes #51351.

## Change

- New `RoutedExpertActivation::SituGlu`, selected at the op boundary and baked into the
  compute kernel as a `SITU_GLU` define, so it caches as its own program.
- SwiGLU-OAI and SiTU-GLU differ only in the SFPU op they invoke, so the kernel's whole
  phase-3 path (keep the raw bf16 gate/up accumulators, skip the gate silu, stream
  gate+up through dst at 2 slots per output tile) is now shared behind `FUSED_BINARY_ACT`
  rather than duplicated. `swiglu_oai_activation_phase` becomes `binary_activation_phase`.
- Betas are baked from `SituGluConfigKimi` (beta_gate=4.0, beta_up=25.0); no extra params.
- Blackhole-only, matching the scoping of `ttnn::softcap` / `ttnn::situ_glu`.

## Dimensions

K3 down-projects the embedding before the routed experts, so the op's K axis is the
post-projection latent 3584, not 7168. Weights are shaped against that: gate/up are
(3584, 3072), down is (3072, 3584). The projection itself is a separate op.

## Performance

Blackhole P150, single chip / single expert, ROW_MAJOR x (production fused-tilize path),
bf8 activations / bf4 weights, allocated == active tokens.
`UnifiedRoutedExpertFfnDeviceOperation` device time; both columns from the same build.

DRAM utilization is against the 512 GB/s spec peak (GDDR6 at 16 Gbps/pin, 8 x 32-bit
channels; tt-smi reports DDR_SPEED=0x3e80=16000).

Core utilization is the standard analytic figure, counting the three matmuls and ignoring
the eltwise/activation work (a good approximation at these shapes):

    FLOP  = 2 * M * K * N * 3 matmuls
    FLOPs = 4096/cycle * 1.35 GHz / fidelity * num_cores     (fidelity = 1, LoFi)
    util  = (FLOP / FLOPs) / measured

against the 88 cores the op uses, for both activations. The FLOP count is identical either
way, so the two utilization columns differ only by measured time; DRAM and TFLOP/s columns are
SiTU-GLU's.

| tokens | SiTU-GLU us | SiLU us | SiTU-GLU util % | SiLU util % | DRAM GB/s | DRAM util % | TFLOP/s |
|---:|---:|---:|---:|---:|---:|---:|---:|
| 32 | 184.72 | 174.27 | 2.4 | 2.5 | 102.5 | 20.0 | 11.4 |
| 64 | 185.26 | 174.56 | 4.7 | 5.0 | 104.1 | 20.3 | 22.8 |
| 128 | 186.18 | 174.80 | 9.3 | 9.9 | 107.3 | 21.0 | 45.4 |
| 256 | 188.26 | 177.57 | 18.5 | 19.6 | 113.6 | 22.2 | 89.8 |
| 512 | 254.14 | 232.81 | 27.4 | 29.9 | 95.2 | 18.6 | 133.1 |
| 1024 | 386.46 | 338.17 | 36.0 | 41.1 | 77.2 | 15.1 | 175.0 |
| 2048 | 680.59 | 595.25 | 40.9 | 46.7 | 60.3 | 11.8 | 198.8 |
| 5120 | 1727.94 | 1513.18 | 40.2 | 45.9 | 43.3 | 8.5 | 195.7 |

DRAM MB per op is 18.58 MB of weights plus activations; all 88 cores participate at every
token count.

Utilization plateaus at ~41% for SiTU-GLU and ~47% for SiLU, both well short of the ~79% a
compute-bound kernel reaches, which is the expected signature of an I/O-bound op:
below the ~256-token knee the kernel is pinned to its DRAM weight-read floor (~185 us
regardless of token count) with cores 2-18% busy, so the extra SFPU work hides in the read
and costs ~6%. Past the knee compute grows into the fixed read, the activation is exposed,
and the cost settles at ~14% -- visible as the ~6-point utilization gap that only opens past
1024 tokens. The lever this leaves is the DRAM weight read itself, which this PR does not
touch.

For reference, DSV3 measured 57% at 5120 tokens under the same formula; K3 sits lower at
40% because its 3584 latent halves the arithmetic intensity per token against a weight read
that shrinks less.

For context, the same measurement against the already-landed SwiGLU-OAI activation at
these dims:

| tokens | SiLU us | SwiGLU-OAI us | SiTU-GLU us | OAI vs SiLU | SiTU-GLU vs SiLU |
|---:|---:|---:|---:|---:|---:|
| 512 | 232.92 | 244.41 | 254.25 | +4.9% | +9.2% |
| 2048 | 594.09 | 639.92 | 680.55 | +7.7% | +14.6% |
| 5120 | 1516.04 | 1628.97 | 1729.17 | +7.5% | +14.1% |

SiTU-GLU costs roughly 2x SwiGLU-OAI, which is what the math predicts: two tanh polynomial
evaluations plus a sigmoid (exp + reciprocal), against OAI's clamp plus one sigmoid.

## Tests

`test_single_routed_expert.py`, Blackhole:

- `test_single_routed_expert_k3_sweep` -- 32/64/128/256/512/1024/2048/5120 tokens, both x
  layouts, at 3584x3072. 16 cases.
- `test_single_routed_expert_k3_saturated` -- drives both tanh caps (sat_deep reaches ~87%
  of gate and ~30% of up past their beta), across bf8 and the production bf4 weights.
- `test_routed_expert_bias.py::test_situglu_bias_routed_expert` -- expert biases + SiTU-GLU.

On the saturation cases: bf8 isolates the activation and holds ~0.999 no matter how deep
the saturation, so a regression in either cap shows up against the 0.97 bar. bf4 is kept
too, at a bar matched to its measured floor (0.976 at scale 0.15, 0.968 at 0.4), because
that cost is real and is what K3 prefill will see. Saturation makes the output near
two-level, so bf4 weight error stops passing through proportionally and instead flips whole
terms of the down matmul -- a property of bf4 plus saturation, not of the kernel (the SFPU
op passes its own saturation-coverage test, and bf8/bf16 hold 0.999 at every scale).

Bias support was previously gated on the SwiGluOai enum value. The kernel's bias branch now
lives in the shared binary-activation phase and is activation-agnostic, so the gate is keyed
to the set of fused binary activations, with the SiTU-GLU half covered by the new test
rather than left permitted-but-unmeasured.

Regression: full `test_swigluoai_routed_expert.py` (7), `test_routed_expert_bias.py`, and
the existing `test_single_routed_expert` functional + ISL sweeps (74) all pass, confirming
the shared phase-3 refactor left the SiLU and SwiGLU-OAI paths intact.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:kgrujcic/51351-k3-re)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:kgrujcic/51351-k3-re)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/unified_routed_expert_ffn/device/kernels/compute/fused_swiglu.cpp [hunk 1]
```diff
@@ -838,9 +869,10 @@ void kernel_main() {
     // gate-intermed write. silu_tile_init() configures the MATH-side SFPU
     // for silu; the pack then runs plain (no per-tile SFPU on the pack
     // thread). Same total compute, better pipelining.
-#ifdef SWIGLU_OAI
-    // SwiGLU-OAI uses the binary swiglu SFPU op (sigmoid/recip table init).
-    MATH((ckernel::llk_math_eltwise_binary_sfpu_swiglu_init()));
+#ifdef FUSED_BINARY_ACT
+    // The binary activations init their own SFPU tables (recip for SwiGLU-OAI,
+    // tanh for SiTU-GLU) instead of silu's.
+    MATH((BINARY_ACT_INIT()));
```
[the remark this answers is not in the document]
Comment (kgrujcic-tt) [reply ttnn/cpp/ttnn/operations/experimental/deepseek_prefill/unified_routed_expert_ffn/device/kernels/compute/fused_swiglu.cpp:861]: Fixed in 725c71519dd. Good catch, and it is the right call for this kernel specifically.

The ordering was pre-existing (both `silu_tile_init()` and the SwiGLU-OAI init already ran before startup), but SiTU-GLU is the case where it actually bites: its `tanh_init` loads the three `vConstFloatPrgm` registers that the activation reads back on every tile, so a startup MMIO write racing those constants would silently corrupt the tanh halves rather than fail loudly.

I moved the activation init to just after `compute_kernel_hw_startup` rather than relocating startup to the top of `kernel_main`: that satisfies the documented contract (startup before any other Compute API init) with a minimal diff, and leaves the count/mailbox handshake that precedes it untouched, since none of that is a Compute API call.

Re-verified after the reorder: all 34 K3 + bias + SwiGLU-OAI cases and the 74 existing functional/ISL cases pass, and device time is unchanged within run-to-run noise (largest delta +1.7% at 1024 tokens, the rest within 0.6%).

**Review by pmilojevicTT — APPROVED** [review]

**Comment by kgrujcic-tt** [comment]
Heads-up for this branch: #52448 moved the fused SFPU code out of the ttnn kernel tree (review feedback). After it merges, `fused_swiglu.cpp` here needs:

- `#include "api/compute/situ_glu.h"` instead of `ttnn/cpp/ttnn/operations/eltwise/binary/device/kernels/compute/situ_glu_sfpu.h`
- `BINARY_ACT_INIT()` -> `situ_glu_tile_init()`, `BINARY_ACT_TILE(fp32, g, u, o)` -> `situ_glu_tile(g, u, o)` (fp32-dest now comes from `DST_ACCUM_MODE`)

**Comment by kgrujcic-tt** [comment]
Superseded by #53602, which is the same work rebased onto current main now that #52448 has landed. Closing this one.

The rebase is not a fast-forward: main gained a full `KimiK3Config` and a SiTU torch reference (#52453's LatentMoE bring-up), so this branch's own copies of both are dropped in favour of main's, and the fused SFPU op moved to `api/compute/situ_glu.h`. #53602 also carries the review feedback from here and runs the K3 MoE end to end on SiTU-GLU.
